rs/http_utils/src/file_downloader.rs

```diff
@@ -472,4 +520,83 @@ mod tests {
             "File already exists, but hash check failed",
         );
     }
+
+    fn create_tar<W: Write>(writer: W) -> io::Result<()> {
+        let mut tar = Builder::new(writer);
+        let mut header = tar::Header::new_gnu();
+        header.set_path("test.txt")?;
+        header.set_size("Hello, world!".as_bytes().len() as u64);
+        header.set_cksum();
+        tar.append(&header, "Hello, world!".as_bytes())?;
+        tar.finish()?;
+        Ok(())
+    }
+
+    #[test]
+    async fn test_is_gz_file() {
+        let temp_dir = tempdir().unwrap();
+        let file_path = temp_dir.path().join("test.tar.gz");
+
+        let tar_gz = File::create(&file_path).unwrap();
+        let mut encoder = GzEncoder::new(tar_gz, Compression::default());
+        create_tar(&mut encoder).unwrap();
+        encoder.finish().unwrap();
+
+        let mut file = File::open(&file_path).unwrap();
+        assert!(is_gz_file(&mut file).unwrap());
+    }
+
+    #[test]
+    async fn test_is_zst_file() {
+        let temp_dir = tempdir().unwrap();
+        let file_path = temp_dir.path().join("test.tar.zst");
+
+        let tar_zst = File::create(&file_path).unwrap();
+        let mut encoder = ZstdEncoder::new(tar_zst, 0).unwrap();
+        create_tar(&mut encoder).unwrap();
+        encoder.finish().unwrap();
+
+        let mut file = File::open(&file_path).unwrap();
+        assert!(is_zst_file(&mut file).unwrap());
+    }
+
+    #[test]
+    async fn test_extract_tar_into_dir_gz() {
+        let temp_dir = tempdir().unwrap();
+        let tar_path = temp_dir.path().join("test.tar.gz");
+        let extract_dir = temp_dir.path().join("extract");
+
+        let tar_gz = File::create(&tar_path).unwrap();
+        let mut encoder = GzEncoder::new(tar_gz, Compression::default());
+        create_tar(&mut encoder).unwrap();
+        encoder.finish().unwrap();
+
+        extract_tar_into_dir(&tar_path, &extract_dir).unwrap();
+
+        let extracted_file = extract_dir.join("test.txt");
+        let mut file = File::open(extracted_file).unwrap();
+        let mut contents = String::new();
+        file.read_to_string(&mut contents).unwrap();
+        assert_eq!(contents, "Hello, world!");
+    }
+
+    #[test]
+    async fn test_extract_tar_into_dir_zst() {
+        let temp_dir = tempdir().unwrap();
+        let tar_path = temp_dir.path().join("test.tar.zst");
+        let extract_dir = temp_dir.path().join("extract");
+
+        let tar_zst = File::create(&tar_path).unwrap();
+        let mut encoder = ZstdEncoder::new(tar_zst, 0).unwrap();
+        create_tar(&mut encoder).unwrap();
+        encoder.finish().unwrap();
+
+        extract_tar_into_dir(&tar_path, &extract_dir).unwrap();
+
+        let extracted_file = extract_dir.join("test.txt");
+        let mut file = File::open(extracted_file).unwrap();
+        let mut contents = String::new();
+        file.read_to_string(&mut contents).unwrap();
+        assert_eq!(contents, "Hello, world!");
+    }
```

Comment:
Consider adding a test case for the error branch.

Reply:
Added! Thank you!